RFC for "Passed Directly" Customization Point (#1999)
uxlfoundation/oneDPL: rfcs/proposed/iterators_pass_directly_customization/README.md

`std::true_type` or `std::false_type` to indicate whether the iterator type `T` should be passed directly to sycl
kernels. There exists an unofficial legacy `is_passed_directly` trait which types can define like this:
`using is_passed_directly = std::true_type;` which is supported within oneDPL. This method is currently used for a
number of helper types within the SYCLomatic compatibility headers, (`device_pointer`, `device_iterator`,
`tagged_pointer`, `constant_iterator`, `iterator_adaptor`). There is no official public API for users who want to
create their own iterator types which could be passed directly to SYCL kernels, this is a gap that should be filled 
with an official public API.

Without something like this users are forced to only rely upon our provided iterator types, or reach into implementation
details which are not part of oneDPL's specified interface.

## Proposal

Create a customization point `oneapi::dpl::is_passed_directly_in_onedpl_device_policies` free function which allows
users to mark their types as passed directly:

```
template <typename T>
constexpr bool is_passed_directly_in_onedpl_device_policies(const T&);
```

> oneDPL will provide a default implementation which will defer to the existing trait:

Reviewer: What header should a user include to get the default implementation? Execution? I think it would be good to specify it here.

Author: Its a good question, and I think I agree that execution is the appropriate place.

Reviewer 2: `<oneapi/dpl/traits>`?

Author: Now that the RFC has evolved a bit to make the public API a trait rather than directly calling the customization point, I think `oneapi/dpl/type_traits` does make more sense than `oneapi/dpl/execution`.

Reviewer: `oneapi/dpl/type_traits` makes sense to me as well.

Reviewer 3: I think it should better go to `<oneapi/dpl/iterator>`.

`type_traits` is OK in principle, but, as far as I can tell, this thing is closer semantically to a concept representing an iterator category (like `forward_iterator`) or a requirement (like `sortable`) than to a generic type trait.

Author: I don't have a strong opinion between these two options.  You are correct that it specifically will augment a custom iterator type, so I don't object to `oneapi/dpl/iterator`.  The inclusion there may perhaps also help users find some combination of our provided custom iterators which may serve their purposes without requiring their own customizations in the first place.

Author: I've added an open question about where this should live.  We can discuss in the soon to come specification PR.

```
template <typename T>
constexpr
bool
is_passed_directly_in_onedpl_device_policies(const T&)
{
	return oneapi::dpl::__ranges::is_passed_directly_v<T>;
}
```

Below is a simple example of a type and customization point definition which is always passed directly.

```
namespace user
{

    struct my_passed_directly_type
    {
        /* unspecified user definition */
    };

    constexpr
    bool
    is_passed_directly_in_onedpl_device_policies(const my_passed_directly_type&)
    {
        return true;
    }
} //namespace user
```

Users can use any constexpr logic based on their type to determine if the type can be passed directly into a SYCL kernel
without any processing. Below is an example of a type which contains a pair of iterators, and should be treated as
passed directly if and only if both base iterators are also passed directly. oneDPL will use this customization point
internally when determining how to handle incoming data, picking up any user defined customizations.

> When using device policies, oneDPL will check iterator types by calling `is_passed_directly_in_onedpl_device_policies`

Reviewer 3: Is it supposed to be a compile-time evaluation or a run-time function call?

Author: As I experiment a little more, I'm realizing that perhaps we can't combine the customization point with a compile-time evaluation requirement, or some other requirements on the iterator types themselves like default constructible.

I'm running into issues attempting to use it within `enable_if` / SFINAE without some way to take the type alone and turn it into a const lvalue ref argument for ADL to work.  If I can call a default constructor, it works, but I'm not sure what other option there is.  

At first I thought `std::declval` was the answer, but it is not.  It is, in fact, both not `constexpr`, and it would be being used in an evaluated expression here. I'll continue experimenting...

Author: I think its only a roadblock for using it in SFINAE, and not in general for `constexpr`. I believe we should be able to refactor our code to lean instead on `if constexr` for "passed directly" checks.

Author: I found a way to deal use it with SFINAE as well, it required returning `std::true_type{}` or `std::false_type` rather than `bool`.  This seems like the right way to go. I've updated the PR accordingly.

# Passed Directly Customization Point for User Defined Types

## Introduction

> oneDPL handles some types of input data automatically as input to its device backend (SYCL-based) as described

Reviewer 4: If I were a user (not oneDPL developer) I would had not understood what phrase "handles some types of input data automatically as input to its device backend" mean... And also I would had not understood what is "passed directly" and why it might be helpful.
I think we have to  explain it more clear by saying about copying data (under an iterator) between the host and device in implicit way in case of "passed directly" is false.

Author: Take a look and see if the changes help clarify this section.

number of helper types within the SYCLomatic compatibility headers (`device_pointer`, `device_iterator`,
`tagged_pointer`, `constant_iterator`, `iterator_adaptor`). There is no official public API for users who want to
create their own iterator types that could be passed directly to SYCL kernels. This is a gap that should be filled 
with an official public API.

Without something like this, users are forced to rely only upon our provided iterator types or reach into implementation
details that are not part of oneDPL's specified interface.

## Proposal
### High Level Proposal

We will create an Argument-Dependent Lookup (ADL) customization point which defines if iterators are "passed directly"
`is_passed_directly_in_onedpl_device_policies`. Users may define if their iterator types should be "passed directly" by
oneDPL by defining a constexpr function in the same namespace where the iterator is defined.
`is_passed_directly_in_onedpl_device_policies` must accept a const lvalue reference to the iterator type being
specialized and return `std::true_type{}` if the iterator type is "passed directly" and `std::false_type{}` otherwise.

Additionally, oneDPL will provide a public trait,

> `inline constexpr bool oneapi::dpl::is_passed_directly_in_onedpl_device_policies_v<T>`, indicating if the iterator type

Reviewer 4: 1) `oneapi::dpl::is_passed_directly_in_onedpl_device_policies_v` has double  info about "oneDPL":
I mean "oneapi::dpl" and substring "_onedpl_".
2) A suffix "_device_policies" is too long.... I would vote to omit that.
So, I would suggest name `oneapi::dpl::is_passed_directly_v`

Author: For now, I think the idea is to merge this RFC with the naming and location as an open question, to be decided in the discussion of the specification PR.  All the same, thanks for your input here.

Also some context for the individual points:
1) The idea behind this is that the public trait follows the naming convention of the ADL customization point function.  The ADL customization point function needs to identify itself as `onedpl` because it user's customizations will be in their own namespaces.  However, we could decide not to have the public trait share naming with the ADL function, and I would agree with removing this redundancy then.  

2) There was an earlier discussion in this PR https://github.com/uxlfoundation/oneDPL/pull/1999#discussion_r1913701632 , I think its a difficult decision with many opinions.  I don't really have strong opinions here and am happy to go with the group's decision.

Author: >Are we also going to introduce a struct as a trait?
 
Its a good question, and probably needs to be handled in connection with the question of naming, and if the ADL function should be named the same as the trait.  

Now I am thinking that it would be better to make their names distinct, and then to offer both the struct and value version of the trait
`oneapi::dpl::is_passed_directly_in_onedpl_device_policies[_v]` as you describe.  This is the more normal convention, but requires divorcing the names between the ADL customization point and the trait.

the iterator should be "passed directly". It would need to pass through all operations to the wrapped base iterator and
make sure no overhead is added in its usage.

There is some complexity in adding such a wrapper iterator, and it would need to be considered carefully to make sure no
 problems would be introduced. This wrapper class may obfuscate users' types and make them more unwieldy to use. It is
 also less expressive than the other options in that it only has the ability to unilaterally mark a type as "passed
 directly". There is no logic that can be used to express some iterator type which may be conditionally "passed
 directly", other than to have logic to conditionally apply the wrapper in the first place. This option seems less clear
 and has more opportunity to cause problems.

## Testing
We will need a detailed test checking both positive and negative responses to
`oneapi::dpl::is_passed_directly_in_onedpl_device_policies_v` to ensure they have the expected result. This should
include tests with custom types, combinations of iterators, USM pointers, etc.

## Open Questions

> * Is there a better, more concise name than `oneapi::dpl::is_passed_directly_in_onedpl_device_policies[_v]` that
> * properly conveys the meaning to the users?

Reviewer 5: ```suggestion
* Is there a better, more concise name than `oneapi::dpl::is_passed_directly_in_onedpl_device_policies[_v]` that
  properly conveys the meaning to the users?
```

Author: thanks, fixed.

make sure no overhead is added in its usage.

There is some complexity in adding such a wrapper iterator, and it would need to be considered carefully to make sure no
 problems would be introduced. This wrapper class may obfuscate users' types and make them more unwieldy to use. It is
 also less expressive than the other options in that it only has the ability to unilaterally mark a type as "passed
 directly". There is no logic that can be used to express some iterator type which may be conditionally "passed
 directly", other than to have logic to conditionally apply the wrapper in the first place. This option seems less clear
 and has more opportunity to cause problems.

## Testing
We will need a detailed test checking both positive and negative responses to
`oneapi::dpl::is_passed_directly_in_onedpl_device_policies_v` to ensure they have the expected result. This should
include tests with custom types, combinations of iterators, USM pointers, etc.

## Open Questions

* Is there a better, more concise name than `oneapi::dpl::is_passed_directly_in_onedpl_device_policies[_v]` that
properly conveys the meaning to the users?

>     * Other names proposed:

Reviewer 4: `oneapi::dpl::is_passed_directly_to_device` ?

Author: I added the two proposed names you raised and a note to the open questions section.

Author: I've separated the names of the ADL customization point and the public trait.  This allows us to provide both the struct and value version of the trait, which is the norm.

* `permutation_iterator` is "passed directly" if both its source iterator and its index map are "passed directly"

### Implementation Details

When using device policies, oneDPL will run compile-time checks on argument iterator types by calling
`is_passed_directly_in_onedpl_device_policies`. If `std::true_type{}` is returned, oneDPL will pass the iterator
directly to SYCL kernels rather than copying the data into `sycl::buffers` and using those buffers to transfer data to
SYCL kernels.

The specification and implementation will be prepared once this RFC is accepted as "proposed". We do not intend to offer
this first as experimental. This RFC will target "Supported" once the specification and implementation are accepted.

### Examples

Below is a simple example of an iterator and ADL customization point definition which is always "passed directly".

```
namespace user
{

    struct my_passed_directly_iterator
    {
        /* unspecified user definition */
    };

    constexpr
    auto
    is_passed_directly_in_onedpl_device_policies(const my_passed_directly_iterator&)
    {
        return std::true_type{};

>     }

Reviewer 3: I would consider this not a typical example. Typically there should be some logic, like in the next example.

>     }

Author: Thinking about the way we use it for our own internal iterator types, I think normally there will be some logic, but there are also some iterator types which have signatures like this.  (`counting_iterator`, `discard_iterator`)

} //namespace user
```

> Users can use any `constexpr` logic based on their iterator to determine if the iterator can be passed directly into a

Reviewer 3: "Users can use any logic" it says.

It is a requirement that the function is `constexpr`, and it is stated earlier. Remember that this is not a formal specification but design description.

The proposal says "When using device policies, oneDPL will run **compile-time checks on argument iterator types** by calling is_passed_directly_in_onedpl_device_policies.". It might be not 100% accurate but the intent is clear (at least to me).

> Users can use any `constexpr` logic based on their iterator to determine if the iterator can be passed directly into a

Author: I've improved the language a bit to make it clear that any logic may be used, as long as it is `constexpr` logic.

> Users can use any `constexpr` logic based on their iterator to determine if the iterator can be passed directly into a

Reviewer 2: In our private discussion with Dan we figured out that we are not talking about `constexpr` function but we are talking about necessity to get the result at compile time. So, a function return type is always a compile-time information, so we should be fine even if users don't put `constexpr` on their customizations. I trust Dan can improve the wording

> Users can use any `constexpr` logic based on their iterator to determine if the iterator can be passed directly into a

Author: I adjusted the language here to remove mention of `constexpr` as that is a red herring but rather focus on defining at compile time the return type.

SYCL kernel without any processing. Below is an example of a type which contains a pair of iterators and should be 
treated as passed directly if and only if both base iterators are also passed directly. Note that the use of the public
trait enables the use of the base iterator type alone without creating a named instance of the base iterator to pass
into the ADL.

```
namespace user
{
    template <typename It1, typename It2>
    struct iterator_pair
    {
        It1 first;
        It2 second;
    };

    template <typename It1, typename It2>
    constexpr 
    auto is_passed_directly_in_onedpl_device_policies(const iterator_pair<It1, It2>&)
    {
        if constexpr (oneapi::dpl::is_passed_directly_in_onedpl_device_policies_v<It1> &&
                      oneapi::dpl::is_passed_directly_in_onedpl_device_policies_v<It2>)
            return std::true_type{};
        else 
            return std::false_type{}; 
    }

> } //namespace user

Reviewer 3: It is more concise but not really simpler. Brevity does not equal to clarity. I personally think that the "decltype" variant is harder to comprehend (and so to write properly).

> } //namespace user

Author: While I'm not sure that your alternative is "simpler", I've included it as another example for those who do. 

I prefer the first version for readability, however I may be missing some advantage the second provides technically.

> } //namespace user

Reviewer 2: When I said "simpler", I referred to the first example that I've provided. I called it "simpler" because it's "logically simpler" in my opinion. In other words, doesn't have redundant branches. In the original RFC text if was written in the style:
```cpp
if (condition)
    return true;
else
    return false;
```
But it seems that you, guys, disagree with me. You might rewrite it with `std::conjuction` if it makes it clearer...

Speaking about a `decltype`: The hole point is when we allow `decltype` scenario we don't force users to write a function definition when having a declaration is sufficient. Some (including myself) would choose to write

> } //namespace user

Author: So, this has evolved a little.  
In an earlier iteration, we only had the `_v` boolean value typed public trait, not the `std::true_type`/`std::false type` public trait as well.  There are definitely places in the draft implementation where it makes more sense to `return condition` as you mention.  

I can update the RFC to reflect this better, and mention usage in "hidden friends" and/or "no body" functions.

> } //namespace user

Author: I've adjusted the examples as mentioned.

method is used for helper types in SYCLomatic compatibility headers (`device_pointer`, `device_iterator`,
`tagged_pointer`, `constant_iterator`, `iterator_adaptor`). However, there is no official public API for users to define
custom iterator types that can be passed directly to SYCL kernels.

This gap should be addressed with an official public API. Without it, users must rely on oneDPL's provided iterator
types or access implementation details outside the specified interface.

## Proposal
### High Level Proposal

We will create an Argument-Dependent Lookup (ADL) customization point which defines if iterators are "passed directly"
`is_passed_directly_in_onedpl_device_policies`. Users may define if their iterator types should be "passed directly" by
oneDPL by defining a `constexpr` function in the same namespace where the iterator is defined.
`is_passed_directly_in_onedpl_device_policies` must accept a const lvalue reference to the iterator type being
specialized and return `std::true_type{}` if the iterator type is "passed directly" and `std::false_type{}` otherwise.

Additionally, oneDPL will provide a public trait,
```cpp
namespace oneapi
{
namespace dpl
{
template <typename T>

> struct is_passed_directly_to_device; // std::true_type or std::false_type

Author: I've adjusted this to adopt close to the language of the specification.

> struct is_passed_directly_to_device; // std::true_type or std::false_type

Reviewer 2: Don't mind about this wording. It's perhaps even better compared to what I proposed. I just wanted to make sure that we don't create a wrong impression that is_passed_directly **is** a `true_type` or `false_type` (because it's impossible in C++ with structs and possible with aliases, which is not our case)

return type based upon the argument iterator type passed.

Additionally, oneDPL will provide a public trait, `is_passed_directly_to_device[_v]` as follows:
```cpp
namespace oneapi
{
namespace dpl
{
template <typename T>
struct is_passed_directly_to_device; // must have the characteristics of std::true_type or of std::false_type

template <typename T>
inline constexpr bool is_passed_directly_to_device_v = oneapi::dpl::is_passed_directly_to_device::value;
} // dpl
} // oneapi
```
indicating if the iterator type `T` is "passed directly". This public trait is intended to be used to help define
`is_passed_directly_in_onedpl_device_policies` for wrapper iterator types which depend on the "passed directly" status
of their base iterator(s) using only the base iterator type(s) rather than a named instance. It may also be used to
confirm user iterator types "passed directly" traits are as intended to prevent unnecessary overhead in oneDPL calls.

### Additional Information

The default implementation of the customization point `is_passed_directly_in_onedpl_device_policies` will be used to
explicitly mark the following iterators as "passed directly":
* Pointers (assumes USM pointers)

> * Iterators containing the legacy `using is_passed_directly = true` trait defined within the type definition
> * Iterators to USM shared allocated `std::vectors` (when knowable)

Reviewer 2: ```suggestion
* Iterators containing the legacy `using is_passed_directly = std::true_type` trait defined within the type definition
* Iterators to USM shared allocated `std::vector`-s (when knowable)
```

    {
        return std::true_type{};
    }
} //namespace user
```

Users can use any compile-time logic based on their iterator type to determine if it can be "passed directly" to define
the ADL function's return type. Commonly, this will involve some check of base template types and their "passed
directly" trait status. Below is an example of a type that contains a pair of iterators and should be treated as "passed
directly" if and only if both base iterators are also "passed directly". Note that the use of the public trait enables
the use of the base iterator type alone without creating a named instance of the base iterator to pass into the ADL.

```cpp
namespace user
{
    template <typename It1, typename It2>
    struct iterator_pair
    {
        It1 first;
        It2 second;
    };

    template <typename It1, typename It2>
    auto is_passed_directly_in_onedpl_device_policies(const iterator_pair<It1, It2>&)
    {
        return std::conjunction<oneapi::dpl::is_passed_directly_to_device<It1>,
                                oneapi::dpl::is_passed_directly_to_device<It2>>{};
    }
} //namespace user
```

> It is also possible to write overloads without a body using an `auto` trailing return type. The following is an
> alternative written in this way for the previous example:

Reviewer 2: It does not matter whether you use a trailing return type of not. It will still work. So, the suggestion is
```suggestion
It is also possible to write overloads without a body. The following is an
alternative written in this way for the previous example:
```

alternative written in this way for the previous example:

```cpp
    template <typename It1, typename It2>

    auto is_passed_directly_in_onedpl_device_policies(const iterator_pair<It1, It2>&) ->
        std::conjunction<oneapi::dpl::is_passed_directly_to_device<It1>,
                         oneapi::dpl::is_passed_directly_to_device<It2>>>;
```
Finally, it is possible to write overloads using the "hidden friend" idiom as functions with or without a body inside
the scope of the iterator definition itself. This option may be preferred when a user wants to ensure that this "passed
directly" trait of their iterator is coupled tightly with the definition itself for maintenance. It also has the
advantage of only being making available for ADL the combinations of template arguments for this iterator which are
explicitly instantiated in the code already.

```cpp
    template <typename It1, typename It2>
    struct iterator_pair
    {
        It1 first;
        It2 second;
        auto is_passed_directly_in_onedpl_device_policies(const iterator_pair&) ->
            std::conjunction<oneapi::dpl::is_passed_directly_to_device<It1>,

>                              oneapi::dpl::is_passed_directly_to_device<It2>>;

Reviewer 3: This is not a "hidden friend" function but a method.

Author: good catch.  I left out the friend keyword.  I'll fix in a second.
